pkg/solana/chainwriter/transform_registry.go

```diff
@@ -80,35 +90,169 @@ func CCIPExecuteArgsTransform(ctx context.Context, client client.MultiClient, ar
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("failed to calculate compute unit limit: %w", err)
+	}
+
+	options = append(options, txmutils.SetEstimateComputeUnitLimit(false), txmutils.SetComputeUnitLimit(computeUnits))
+
+	if len(argsTransformed.Info.AbstractReports) != 1 {
+		return nil, nil, nil, nil, fmt.Errorf("encountered unexpected number of reports, got %d, expect 1", len(argsTransformed.Info.AbstractReports))
+	}
+	report := argsTransformed.Info.AbstractReports[0]
+	if len(report.Messages) != 1 {
+		return nil, nil, nil, nil, fmt.Errorf("encountered unexpected number of messages, got %d, expect 1", len(report.Messages))
+	}
+
+	message := report.Messages[0]
+	sourceChainSel := message.Header.SourceChainSelector
+	if len(argsTransformed.Info.MerkleRoots) != 1 {
+		return nil, nil, nil, nil, fmt.Errorf("encountered unexpected number of merkle roots, got %d, expect 1", len(argsTransformed.Info.MerkleRoots))
+	}
+	merkleRoot := argsTransformed.Info.MerkleRoots[0].MerkleRoot
+
+	var messageAccounts []ccip_offramp.CcipAccountMeta
+	if !message.Receiver.IsZeroOrEmpty() {
+		logicReceiver := solana.PublicKeyFromBytes(message.Receiver)
+		// Append logic receiver as the first messaging account for derivation
+		messageAccounts = append(messageAccounts, ccip_offramp.CcipAccountMeta{
+			Pubkey:     logicReceiver,
+			IsSigner:   false,
+			IsWritable: false,
+		})
+		// Extract the user defined accounts
+		userAccountsLookup := AccountLookup{
+			Name:       "UserAccounts",
+			Location:   "ExtraData.ExtraArgsDecoded.accounts",
+			IsWritable: MetaBool{BitmapLocation: "ExtraData.ExtraArgsDecoded.accountIsWritableBitmap"},
+			IsSigner:   MetaBool{Value: false},
+		}
+		userAccounts, resolveErr := userAccountsLookup.Resolve(args)
+		// If err is ErrLookupNotFoundAtLocation, allow process to continue in case accounts are not needed
+		if resolveErr != nil && !errors.Is(resolveErr, ErrLookupNotFoundAtLocation) {
+			return nil, nil, nil, nil, fmt.Errorf("failed to resolve user accounts: %w", resolveErr)
+		}
+		messageAccounts = append(messageAccounts, ConvertToCCIPAccountMetas(userAccounts)...)
+	}
+
+	// Extract token transfers
+	var tokenTransfers []ccip_offramp.TokenTransferAndOffchainData
+	var tokenReceiver solana.PublicKey
+	var messageTokenData [][]byte
+	if len(message.TokenAmounts) > 0 {
+		tokenTransfers = make([]ccip_offramp.TokenTransferAndOffchainData, 0, len(message.TokenAmounts))
+		if len(argsTransformed.ExtraData.DestExecDataDecoded) != len(message.TokenAmounts) {
+			return nil, nil, nil, nil, fmt.Errorf("unexpected number of DestExecData encountered. expect the same number as token transfers %d, got %d", len(message.TokenAmounts), len(argsTransformed.ExtraData.DestExecDataDecoded))
+		}
+		// If message contains token transfers, extract offchain token data for the message if any exists
+		if len(report.OffchainTokenData) != len(report.Messages) {
```

Comment:
nit: this works because you previously confirmed `len(report.Messages) ==1`, but it looks weird here since the check looks like it's to ensure `report.OffchainTokenDat[0]` is valid.
```suggestion
		if len(report.OffchainTokenData) != 1 {
```

Reply:
Ya I was trying to reduce where we'd need to make changes if we did support multiple messages. It might not be obvious that `OffchainTokenData` length was tied to `Messages` length otherwise. But it does look out of place here. I'll make this change and just leave a comment instead